### PR TITLE
Fix android low processor usage mode flicker 3.x

### PR DESCRIPTION
**Pull request by Gromph**
This fixes https://github.com/godotengine/godot/issues/19304
The fix is to draw 2 more frames after VisualServer::has_changed returns false.
The next problem this also fixes is if the gl surface is lost (for example turning screen off/on, app switching, rotating device) then the screen sometimes would be black, or flicker between a rendered frame and a black frame. The fix for this is to draw 3 frames after any gl surface changed notifications from android

Master does not seem to have the original flicker problem, but the screen does stay black after app switching or screen off/on. So submitting a pull request for 3.x for now.

### CODE REVIEW

**Comment by Calinou** [comment]
I tested this PR on a OnePlus 6 running Android 11 and it seems to work! Previously, I would get a lot of flickering and a black screen once returning to the app. Both issues are resolved by this PR :slightly_smiling_face: 

**Testing project:** [test_low_processor_mode.zip](https://github.com/godotengine/godot/files/7990691/test_low_processor_mode.zip)

Android APK with this PR included: [Test Low Processor Mode.zip](https://github.com/godotengine/godot/files/7990785/Test.Low.Processor.Mode.zip)
<sub>(extract the ZIP archive and find the APK inside)</sub>

Android APK exported from 3.4.2 without this PR: [Test Low Processor Mode.zip](https://github.com/godotengine/godot/files/7990670/Test.Low.Processor.Mode.zip)
<sub>(extract the ZIP archive and find the APK inside)</sub>

**Review by Calinou — APPROVED** [review]

**Review by m4gr3d — CHANGES_REQUESTED** [review]
Not sure if I agree with the fix since it looks like we're applying a band-aid instead of fixing the root cause of the issue.

Do we know why drawing a few more frames removes the flicker, or why it's occurring in low-power mode in the first place.
Happy to defer to someone with more graphics expertise.

**Comment by Gromph** [comment]
It is definitely a bandaid, I occasionally still saw flickers. So in our fork of godot I upped the number of frames force drawn from 3 to 6 in  S_Android::set_display_size.    Haven't seen any flickers since.

**Comment by golddotasksquestions** [comment]
>So in our fork of godot I upped the number of frames force drawn from 3 to 6 in S_Android::set_display_size. Haven't seen any flickers since.

Is this a possible solution for everyone? If so can this fix be added to core?
In my current card game project, my phone battery is drained by 18% if I just leave it and there is literally nothing changes during that time. 
So I really need low processing mode for all the time you just stare at the screen and think your next move and it does not need to be refreshed at all. Without it Godot is a really bad choice for these type of mobile games and apps. Unfortunately right now low processing mode fickers so much it's not even usable for testing (on Android).

**Comment by Calinou** [comment]
> Is this a possible solution for everyone? If so can this fix be added to core?

It's pretty hacky (and will not save as much power as doing it properly), so I'd recommend you integrate this to a [custom Android export template build](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_android.html) right now.

**Comment by golddotasksquestions** [comment]
@Calinou Thanks for the suggestion, unfortunately custom builds is still above my current level of technical ability.

Not sure how hacky or stupid this is, but this is my current workaround:

I add a `Timer` node to my main scene, set `wait_time` to something like 5 seconds, and then these short lines of script:

```
func _input(event):
	Engine.target_fps = 0
	low_fps_timer.start()

func _on_low_fps_timer_timeout():
	Engine.target_fps = 1
```
This will reduce the frames drawn to 1 per second and set it back to the default 0 (0 in this case meaning unlimited), whenever there is any input of any kind at all.

With this I was able to reduce the battery drain from 18% to 5% over a period of 30 minutes.

**Comment by Relintai** [comment]
I actually managed to figure it out.

#### TL;DR

The issue is that on the android backend the framebuffers are swapped automatically by android after every frame, whether the engine want it or not. That's why you see flicker, because you see a previous (black if not yet rendered) and the current frame swap constantly.

#### Not TL;DR

The way it's set up right now, is that the [main view](https://github.com/godotengine/godot/blob/3.x/platform/android/java/lib/src/org/godotengine/godot/GodotView.java ) class always just calls [`GodotRenderer`](https://github.com/godotengine/godot/blob/3.x/platform/android/java/lib/src/org/godotengine/godot/GodotRenderer.java)'s `onDrawFrame()`, which in turn calls [`Main::iteration()`](https://github.com/godotengine/godot/blob/3.x/main/main.h), and then it automatically swaps framebuffers. 

However godot's [`Main::iteration()`](https://github.com/godotengine/godot/blob/3.x/main/main.h) is set up in a way where it handles inputs, updates the world, and then it decides whether it should draw or not, and if not, it just keeps the previous framebuffer in place, and then does some cleanups, so it always need to be called.

Note: Swapping buffers is implemented in OS, for example for windows it's [[here]](https://github.com/godotengine/godot/blob/c08e753ef6dba156a87bb4148faf7d190b6676f0/platform/windows/os_windows.cpp#L3406). Android does not have this implemented, as with the current setup it's unavailable anyway due to how the backend is set up.

The simplest fix would be to create a `GLSurfaceView` implementation, that lets the user control framebuffer swaps. However I don't know whether that's possible or not. If it's not possible that means `Main` would need to be changed by potentially quite a bit, along with the android backend.

**Comment by akien-mga** [comment]
Superseded by #59606.

Thanks @Gromph and @Relintai for your work, that was key to develop the fix that got merged.

**Comment by Helioform** [comment]
It crashes everytime I try to make a Windows package. And I dont see any way to make an Android .apk or .aab file since it requires the SDK.

**Comment by Helioform** [comment]
Sorry I should have started another thread. But this was for the Android
apk that I tried. I cannot package anything with it so I was wondering if
this is a bug or just not implemented yet.

Thanks.

**Comment by Calinou** [comment]
Exporting from the Android editor isn't a supported use case yet. It will probably take a while until everything is stable enough to consider the Android editor production-ready.
